### ISSUE TITLE
MAX_STACK_ALLOC + Segfault

### ISSUE THREAD

**Issue by bashimao** [xianyi/OpenBLAS#567]
Changing the default value of MAX_STACK_ALLOC may not have been a good idea. Even with the changes from (https://github.com/jeromerobert/OpenBLAS/commit/b17ccb4c5c5ef3a41b7a5040f1c604348f8cf6f0), I am still getting a Segfaults.

<b>Example:</b>
Multiplication of a 24 x 18 matrix with a 18 x 1 vector (filled with random doubles [0.0, 1.0]).

<b>No problems with:</b>
<code>
make BINARY=64 TARGET=NEHALEM USE_THREAD=0 MAX_STACK_ALLOC=0
</code>
<code>
make BINARY=64 TARGET=SANDYBRIDGE USE_THREAD=0 MAX_STACK_ALLOC=0
</code>
<code>
make BINARY=64 TARGET=HASWELL USE_THREAD=0 MAX_STACK_ALLOC=0
</code>
<code>
make BINARY=64 TARGET=NEHALEM USE_THREAD=0 MAX_STACK_ALLOC=2048
</code>
<code>
make BINARY=64 TARGET=SANDYBRIDGE USE_THREAD=0 MAX_STACK_ALLOC=2048
</code>

<b>Segfault with:</b>
<code>
make BINARY=64 TARGET=HASWELL USE_THREAD=0 MAX_STACK_ALLOC=2048
</code>

Please note that I am also getting segfaults, if I enforce the Sandybridge implementation together with MAX_STACK_ALLOC > 0. It just does not happen for this particular case.

<b>System Detail:</b>
cpu vendor_id : GenuineIntel
cpu family : 6
model : 60
model name : Intel(R) Core(TM) i7-4770 CPU @ 3.40GHz
stepping : 3
microcode : 0x16
OS ... Linux ====> LinuxMint / Ubuntu, Trusty with kernel 3.16.0-36
Architecture ... x86_64

BINARY ... 64bit

C compiler ... GCC (command line : gcc) ====> gcc (Ubuntu 4.8.2-19ubuntu1) 4.8.2
Fortran compiler ... GFORTRAN (command line : gfortran) ===> GNU Fortran (Ubuntu 4.8.2-19ubuntu1) 4.8.2

**Comment by martin-frbg** [comment]
Which version of OpenBLAS is this ? That patch has already been applied in the develop branch.
Also can you share a small code sample that displays the problem, just in case it might be something different ?

**Comment by bashimao** [comment]
I compile from these sources: http://github.com/xianyi/OpenBLAS/archive/v0.2.14.zip
(already contains the "volatile" fix)

Since I use scalanlp/Breeze, which in turn interfaces with BLAS via fommil/netlib-java (JNI binding) it is hard to tell what the actual invocation looks like. If you have these libraries and a Scala REPL you can try the following commands to reproduce the problem.

<pre>
import breeze.linalg._
import scala.util._
val a = DenseMatrix.fill[Double](24, 18)(Random.nextDouble())
val b = DenseVector.fill[Double](18)(Random.nextDouble())
a * b
</pre>

**Comment by zerothi** [comment]
First try the development version.
See for instance:
https://github.com/xianyi/OpenBLAS/commit/a4c96eca679664a70c34b40026bbc5ad55825fd2
Which has been applied later.

2015-05-12 14:04 GMT+02:00 Matthias Langer <notifications@github.com>:

> I compile from these sources:
> http://github.com/xianyi/OpenBLAS/archive/v0.2.14.zip
> <https://github.com/xianyi/OpenBLAS/archive/v0.2.14.zip>
> (already contains the "volatile" fix)
>
> Since I use scalanlp/Breeze, which in turn interfaces with BLAS via
> fommil/netlib-java (JNI binding) it is hard to tell what the actual
> invocation looks like. If you have these libraries and a Scala REPL you can
> try the following commands to reproduce the problem.
>
> import breeze.linalg._
> import scala.util._
> val a = DenseMatrix.fill[Double](24, 18)(Random.nextDouble())
> val b = DenseVector.fill[Double](18)(Random.nextDouble())
> a * b
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/xianyi/OpenBLAS/issues/567#issuecomment-101252668>.
>



-- 
Kind regards Nick

**Comment by bashimao** [comment]
This is probably not your problem, but maybe you can help me here. Compiling the current head-revision of OpenBLAS works fine. However, fommil/netlib-java's JNI loader fails to recognize the library as a valid BLAS implementation. libopenblas_haswell-r0.2.14.so is definitely there and update-alternatives also points to the right file. Not sure why it refuses the library. When I compile the version from the homepage (http://github.com/xianyi/OpenBLAS/archive/v0.2.14.zip) there is no such problem. Did anything significant regarding the library's internal structure change since that release?

**Comment by zerothi** [comment]
I am not experienced at all in java. Does it print any debugging
information?

I haven't experienced anything I should change, but then again, java hasn't
been my environment.

2015-05-12 15:15 GMT+02:00 Matthias Langer <notifications@github.com>:

> This is probably not be your problem, but maybe you can help me here.
> Compiling the current head-revision of OpenBLAS works fine. However,
> fommil/netlib-java's JNI loader fails to recognize the library as a valid
> BLAS implementation. libopenblas_haswell-r0.2.14.so is definitely there
> and update-alternatives definitely points to the right file. Not sure why
> it refuses it. When I compile the version from the homepage (
> http://github.com/xianyi/OpenBLAS/archive/v0.2.14.zip
> <https://github.com/xianyi/OpenBLAS/archive/v0.2.14.zip>) there is no
> such problem. Did anything significant about the library's internal
> structure change since that release?
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/xianyi/OpenBLAS/issues/567#issuecomment-101270757>.
>



-- 
Kind regards Nick

**Comment by xianyi** [comment]
cc @fommil ,
@bashimao  , I think you need contact netlib-java project.

**Comment by fommil** [comment]
If it doesn't work from netlib java then it means you're not meeting the CBLAS / LAPACK API standard. It works for all other implementations I know about. Either the OP has a DevOps issue or your API is broken.

**Comment by bashimao** [comment]
@fommil @xianyi Thanks for your help. Well this is a little bit embarrassing. I am now able to load the library with netlib-java. To be honest, I am not sure what went wrong before. My environment settings have not changed. I also checked the update log of my OS. But there is nothing that can explain this. For what it's worth. The solution to this problem is apparently: <i>"Sleep one night and and reboot the machine!"</i>

@xianyi  @zerothi Having now solved the load-library-issue, I can hereby confirm that this was a bug in the 0.2.14 (d0c51c4) release that has already been solved in the development version. The change that actually solves my particular segfault problem is commit 9798481, which adds the blas_memory_alloc_nolock function.
